pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/SegmentPreprocessThrottler.java

```diff
@@ -0,0 +1,212 @@
+   * @param relaxThrottling whether to relax throttling prior to serving queries
+   */
+  public SegmentPreprocessThrottler(int maxPreprocessConcurrency, int maxPreprocessConcurrencyBeforeServingQueries,
+      boolean relaxThrottling) {
+    LOGGER.info("Initializing SegmentPreprocessThrottler, maxPreprocessConcurrency: {}, "
+            + "maxPreprocessConcurrencyBeforeServingQueries: {}, relaxThrottling: {}",
+        maxPreprocessConcurrency, maxPreprocessConcurrencyBeforeServingQueries, relaxThrottling);
+    Preconditions.checkArgument(maxPreprocessConcurrency > 0,
+        "Max preprocess parallelism must be > 0, but found to be: " + maxPreprocessConcurrency);
+    Preconditions.checkArgument(maxPreprocessConcurrencyBeforeServingQueries > 0,
+        "Max preprocess parallelism before serving queries must be > 0, but found to be: "
+            + maxPreprocessConcurrencyBeforeServingQueries);
+
+    _maxPreprocessConcurrency = maxPreprocessConcurrency;
+    _maxPreprocessConcurrencyBeforeServingQueries = maxPreprocessConcurrencyBeforeServingQueries;
+    _relaxThrottling = relaxThrottling;
+
+    // maxConcurrentPreprocessesBeforeServingQueries is only used prior to serving queries and once the server is
+    // ready to serve queries this is not used again. Thus, it is safe to only pick up this configuration during
+    // server startup. There is no need to allow updates to this via the ZK CLUSTER config handler
+    int relaxThrottlingThreshold = Math.max(_maxPreprocessConcurrency, _maxPreprocessConcurrencyBeforeServingQueries);
+    int preprocessConcurrency = _maxPreprocessConcurrency;
+    if (relaxThrottling) {
+      preprocessConcurrency = relaxThrottlingThreshold;
+      LOGGER.info("Relax throttling enabled, setting preprocess concurrency to: {}", preprocessConcurrency);
+    }
+    _semaphore = new AdjustableSemaphore(preprocessConcurrency, true);
+    LOGGER.info("Created semaphore with total permits: {}, available permits: {}", totalPermits(),
+        availablePermits());
+  }
+
+  public synchronized void resetThrottling() {
+    LOGGER.info("Reset throttling threshold for segment preprocess concurrency, total permits: {}, available "
+            + "permits: {}", totalPermits(), availablePermits());
+    _relaxThrottling = false;
+    _semaphore.setPermits(_maxPreprocessConcurrency);
+    LOGGER.info("Reset throttling completed, new concurrency: {}, total permits: {}, available permits: {}",
+        _maxPreprocessConcurrency, totalPermits(), availablePermits());
+  }
+
+  @Override
+  public synchronized void onChange(Set<String> changedConfigs, Map<String, String> clusterConfigs) {
+    if (clusterConfigs == null || clusterConfigs.isEmpty()) {
+      LOGGER.info("Skip updating SegmentPreprocessThrottler configs with empty clusterConfigs");
+      return;
+    }
+
+    if (changedConfigs == null || changedConfigs.isEmpty()) {
```

Comment:
(minor) This is equivalent to `CollectionUtils.isEmpty(changedConfigs)`

Reply:
done